src/servers/launcher/main.cpp

```diff
@@ -0,0 +1,849 @@
+#include <libpq-fe.h>
+
+#include <array>
+#include <cerrno>
+#include <chrono>
+#include <cstdint>
+#include <cstring>
+#include <expected>
+#include <iomanip>
+#include <iostream>
+#include <netinet/in.h>
+#include <ranges>
+#include <signal.h>
+#include <stdexcept>
+#include <sstream>
+#include <string>
+#include <sys/epoll.h>
+#include <sys/wait.h>
+#include <thread>
+#include <unistd.h>
+#include <utility>
+#include <variant>
+
+using namespace std::literals;
+using notManyMilliseconds = std::chrono::duration<int, std::milli>;
+using PacketBuffer = std::array<uint8_t, 65536>;
+
+static constexpr const auto RETRY_DELAY{1s};
+static constexpr const uint16_t PORT{32232};
+static constexpr const uint8_t CABOOSE{108};
+
+struct UUID {
+  std::array<const uint8_t, 16> uuid;
+
+  UUID() : uuid(std::array<const uint8_t, 16>{
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+  }) {}
+  UUID(std::ranges::sized_range auto&& uuid) :
+    uuid([&uuid]{
+      static_assert(uuid.size() == 16, "UUIDs must be 16 bytes long");
+
+      return std::array<const uint8_t, 16>{
+        uuid[0],
+        uuid[1],
+        uuid[2],
+        uuid[3],
+        uuid[4],
+        uuid[5],
+        uuid[6],
+        uuid[7],
+        uuid[8],
+        uuid[9],
+        uuid[10],
+        uuid[11],
+        uuid[12],
+        uuid[13],
+        uuid[14],
+        uuid[15],
+      };
+    }()) {}
+  static UUID unsafeCreate(const char* v) {
+    // No bounds checking, the caller is responsible for providing the
+    // right length array.
+    std::array<uint8_t, 16> uuid{};
+    for (size_t i{0}; i < 16; i++) uuid[i] = static_cast<uint8_t>(v[i]);
+
+    return UUID{uuid};
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const UUID& uuid) {
+    auto flags{os.flags()};
+    os << std::hex;
+    for (const auto& [i, c] : uuid.uuid | std::views::enumerate) {
+      os << std::setw(2) << std::setfill('0') << +c;
+      if (i == 3 || i == 5 || i == 7 || i == 9) os << '-';
+    }
+    os.flags(flags);
+
+    return os;
+  }
+};
+
+struct Context {
+  const UUID* SERVER_ID;
+};
+
+static Context ctx{};
+
+class DBConnectionError: public std::runtime_error {
+  public:
+  DBConnectionError(const std::string& error) noexcept:
+    std::runtime_error(error) {}
+};
+
+class DB {
+  PGconn* c;
+
+  DB() :
+    c([]{
+      PGconn* c{PQconnectdb("")};
+      if (PQstatus(c) != CONNECTION_OK) {
+        throw DBConnectionError(PQerrorMessage(c));
+      }
+      return c;
+    }()) {}
+
+  public:
+  static std::expected<DB, std::string> tryCreate() noexcept {
+    try {
+      return DB();
+    } catch (const DBConnectionError& exc) {
+      return std::unexpected(exc.what());
+    }
+  }
+  static const DB create() noexcept {
```

Comment:
Returning a `const` value like this is an anti-pattern. Most of the time the `const` will be ignored in any case. And when it’s not, it’s usually a pain in the ass.